Rules/UseCompatibleTypes.cs

```diff
@@ -0,0 +1,854 @@
+            {
+                WriteToLogFile(logFile, "There are no target platforms listed in the compatibility parameter.");
+                return;
+            }
+
+            // Create our platformSpecMap from the target platforms in the compatibilityList.
+            foreach (string target in compatibilityList)
+            {
+                string psedition, psversion, os;
+
+                // ignore invalid entries
+                if (GetVersionInfoFromPlatformString(target, out psedition, out psversion, out os))
+                {
+                    platformSpecMap.Add(target, new { PSEdition = psedition, PSVersion = psversion, OS = os });
+                }
+            }
+
+            // Find corresponding dictionaries for target platforms and create type maps.
+            ProcessDirectory(settingsPath, compatibilityList);
+
+            if (psTypeMap.Keys.Count != compatibilityList.Count())
+            {
+                WriteToLogFile(logFile, "One or more of the target platforms listed in the compatibility parameter is not valid.");
+                return;
+            }
+
+            // Set up the reference type map (from desktop PowerShell).
+            CreateReferenceTypeMap(settingsPath, referenceFileName);
+
+            // Set up type accelerator map.
+            CreateTypeAcceleratorMap(settingsPath);
+
+            // Reached this point, so no initialization error.
+            hasInitializationError = false;
+        }
+
+        /// <summary>
+        /// Search the Settings directory for files in the form [PSEdition]-[PSVersion]-[OS].json.
+        /// For each json file found that matches our target platforms, parse file to create type map.
+        /// </summary>
+        private void ProcessDirectory(string path, List<string> compatibilityList)
+        {
+            var jsonFiles = Directory.EnumerateFiles(path, "*.json");
+            if (jsonFiles == null)
+            {
+                return;
+            }
+
+            foreach (string file in jsonFiles)
+            {
+                string fileNameWithoutExt = Path.GetFileNameWithoutExtension(file);
+                if (!compatibilityList.Contains(fileNameWithoutExt, StringComparer.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                dynamic deserialized = JObject.Parse(File.ReadAllText(file));
+                psTypeMap[fileNameWithoutExt] = GetTypesFromData(deserialized);
+            }
+        }
+
+        /// <summary>
+        /// Get a hashset of full type names from a deserialized json file.
+        /// </summary>
+        private HashSet<string> GetTypesFromData(dynamic deserializedObject)
+        {
+            HashSet<string> types = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            dynamic typeList = deserializedObject.Types;
+            foreach (dynamic type in typeList)
+            {
+                string name = type.Name.ToString();
```

Comment:
These should already be strings -- it might be a case of casting them rather than calling `ToString()`? Unless there's something occuring in deserialisation?